packages/astro/src/core/render-context.ts

```diff
@@ -311,9 +311,12 @@ export class RenderContext {
 			(await pipeline.componentMetadata(routeData)) ?? manifest.componentMetadata;
 		const headers = new Headers({ 'Content-Type': 'text/html' });
 		const partial = typeof this.partial === 'boolean' ? this.partial : Boolean(mod.partial);
+		const actionResult = hasActionPayload(this.locals)
+			? deserializeActionResult(this.locals._actionPayload.actionResult)
```

Comment:
There's a slight chance that `_actionPayload` isn't there, in case locals get cleaned by the user, so we should use a safer approach. 

If we're inside an action, and we must have `_actionPayload`, then we it's safer to:
1. Check if rendering the result of action (`this.routePattern` should be enough)
2. Throw an error if `_actionPayload` isn't there. I think throwing an error is reasonable because users have wiped `locals`, when they should not.

Reply:
Oh hm, I don't think that's possible for actions submitted by a form. `_actionPayload` may be populated after a redirect using cookies or session storage, so we can't rely on the route to determine if `_actionPayload` should exist.

It doesn't look like this PR is changing the behavior here either (unless I'm missing something nuanced). I'd vote to address this in a separate PR if there's a solution. Only thought I have is using `Object.create()` to set the action payload to avoid accidental deletion